src/common/HeartbeatMap.h

```diff
@@ -39,14 +39,17 @@ namespace ceph {
 
 struct heartbeat_handle_d {
   const std::string name;
-  pthread_t thread_id;
-  // TODO: use atomic<time_point>, once we can ditch GCC 4.8
-  std::atomic<unsigned> timeout = { 0 }, suicide_timeout = { 0 };
-  time_t grace, suicide_grace;
+  pthread_t thread_id = 0;
+  using clock = ceph::coarse_mono_clock;
+  using time = ceph::coarse_mono_time;
+  std::atomic<time> timeout = clock::zero();
+  std::atomic<time> suicide_timeout = clock::zero();
```

Comment:
might be worth adding a `static_assert(std::atomic<time>::is_always_lock_free)` in case this gets backported anywhere (upstream or down)

Reply:
why, just to make sure we don't have performance regressions after replacing `unsigned` with `ceph::coarse_mono_time`?

and i would be nervous if we have 
```c++
static_assert(std::atomic<time>::is_always_lock_free);
```
in our code. as it will breaks the build on x86, armhf, mips and mipsel, which are 32-bit architectures. as some of our downstream maintainers are still packaging Ceph for these architectures for the distros they are working on. and AFAIK, some users in our communities are running Ceph on systems with 32-bit ARM CPUs.

because we use a 64bit number to representing a `coarse_mono_clock` time,  on these 32-bit platforms, `std::atomic<time>::is_always_lock_free` is evaluated `false` because libatomic would kick in in this case. but before this change, we use
```c++
std::atomic<unsigned> timeout = { 0 }, suicide_timeout = { 0 };
```
these atomic variables can be accessed without lock, hence `std::atomic<unsigned>::is_always_lock_free` is true. well, as we know, to use a 32-bit number to represent the epoch time in seconds is not encouraged -- Y2038.

Reply:
ok, thanks @tchaikov - to summarize, then, 32-bit architectures will wrap this in a spinlock, but we're okay with a performance hit there because we want a fix for Y2038

Reply:
> 32-bit architectures will wrap this in a spinlock

i am not familiar with how libatomic implements the 64bit store/load operations on these architectures. but after some research, i found that on x86, it's guaranteed that 64bit read/write a quardword on aligned on 64-bit binary is atomic. see https://xem.github.io/minix86/manual/intel-x86-and-64-manual-vol3/o_fe12b1e2a880e0ce-258.html

but on other architectures, i think there is good chance they have exclusive load/store instructions and alike to implement a retry machinery for handling split read/write.